File: typescript/src/AuthManager.ts
import { type AuthProvider, type AuthResult } from './types.js';
import { verifyToken } from './utils/jwt.js';
import { type JWTPayload } from 'jose';

export class AuthManager {
  private providers = new Map<string, AuthProvider>();
  private jwtSecret: string;

  constructor(jwtSecret: string) {
    this.jwtSecret = jwtSecret;
  }

  /** Register a provider under a given name (defaults to provider.name) */
  register(provider: AuthProvider, alias?: string): this {
    const key = alias ?? provider.name;
    this.providers.set(key, provider);
    return this;
  }

  /** Authenticate using a named provider */
  async authenticate(
    providerName: string,
    credentials: Record<string, unknown>,
  ): Promise<AuthResult> {
    const provider = this.providers.get(providerName);
    if (!provider) {
      return { success: false, error: `Unknown provider: ${providerName}` };
    }
    return provider.authenticate(credentials);
  }

  /** Verify a JWT issued by this framework */
  async verifyToken(token: string): Promise<JWTPayload> {
    return verifyToken(token, this.jwtSecret);
  }

  /** List registered provider names */
  listProviders(): string[] {
    return Array.from(this.providers.keys());
  }
}

File: python/src/authy/auth_manager.py
"""AuthManager: registry and orchestrator for all authentication providers."""
from __future__ import annotations

from typing import Any

from .types import AuthProvider, AuthResult
from .utils.jwt_utils import verify_token


class AuthManager:
    """Registers providers and routes authentication requests.

    Example::

        manager = AuthManager(jwt_secret="secret")
        manager.register(LocalProvider(config, find_user))
        result = await manager.authenticate("local", {"username": "alice", "password": "pw"})
    """

    def __init__(self, jwt_secret: str) -> None:
        self._jwt_secret = jwt_secret
        self._providers: dict[str, AuthProvider] = {}

    def register(self, provider: AuthProvider, alias: str | None = None) -> "AuthManager":
        """Register a provider. Returns self for chaining."""
        key = alias or provider.name
        self._providers[key] = provider
        return self

    async def authenticate(
        self, provider_name: str, credentials: dict[str, Any]
    ) -> AuthResult:
        """Authenticate using a named provider."""
        provider = self._providers.get(provider_name)
        if provider is None:
            return AuthResult(success=False, error=f"Unknown provider: {provider_name!r}")
        return await provider.authenticate(credentials)

    def verify_token(self, token: str) -> dict[str, Any]:
        """Verify a JWT issued by this framework. Raises jwt.InvalidTokenError on failure."""
        return verify_token(token, self._jwt_secret)

    def list_providers(self) -> list[str]:
        """Return a list of registered provider names."""
        return list(self._providers.keys())
